feat: DNSSEC Validation

developStorm opened the pull request:
resolves #441

# Purpose
Adds support for validating DNSSEC records up to the root's chain-of-trust

The level of detail in the validation results depends on the `--result-verbosity=` flag. In the `short` mode, validation results are not output (equivalent to no validation). In the `normal` mode, only the overall validation status is output. In the `long` mode, all relevant records used for validating this specific result (DS, DNSKEY, RRSIG) and the validation results for each individual RRset are printed. In the `trace` mode, all validation details for the entire query chain are available along the trace.

Since we are a DNS scanning tool, the resolution results may still hold value even in cases of DNSSEC validation failure. Therefore, a validation failure does not lead to a resolution failure - i.e., the resolution results will still be output normally. It is the responsibility of data consumers to check the DNSSEC validation status before using the results.

Example usage:
```shell
go run . A example.com --iterative --validate-dnssec --result-verbosity=long
```

Example output (long):
```json
{
    "class": "IN",
    "name": "example.com",
    "results": {
        "A": {
            "data": {
                "answers": [
                    {
                        "answer": "93.184.215.14",
                        "class": "IN",
                        "name": "example.com",
                        "ttl": 3600,
                        "type": "A"
                    },
                    {
                        "algorithm": 13,
                        "class": "IN",
                        "expiration": "20241206041826",
                        "inception": "20241115121713",
                        "keytag": 60915,
                        "labels": 2,
                        "name": "example.com",
                        "original_ttl": 3600,
                        "signature": "vRP5fmIUOIHdIXJsnFkSLYL84OEh4isvOeIkPdGR1Ro+YUs4tTNO6J3ajnDvcIZfc5p+j8R35Z/SOOqD9ZY95w==",
                        "signer_name": "example.com.",
                        "ttl": 3600,
                        "type": "RRSIG",
                        "type_covered": 1
                    }
                ],
                "dnssec": {
                    "additionals": [
                        {
                            "error": "",
                            "rrset": {
                                "class": 4096,
                                "name": ".",
                                "type": 41
                            },
                            "sig": null,
                            "status": "Insecure"
                        }
                    ],
                    "answer": [
                        {
                            "error": "",
                            "rrset": {
                                "class": 1,
                                "name": "example.com.",
                                "type": 1
                            },
                            "sig": {
                                "algorithm": 13,
                                "class": "IN",
                                "expiration": "20241206041826",
                                "inception": "20241115121713",
                                "keytag": 60915,
                                "labels": 2,
                                "name": "example.com",
                                "original_ttl": 3600,
                                "signature": "vRP5fmIUOIHdIXJsnFkSLYL84OEh4isvOeIkPdGR1Ro+YUs4tTNO6J3ajnDvcIZfc5p+j8R35Z/SOOqD9ZY95w==",
                                "signer_name": "example.com.",
                                "ttl": 3600,
                                "type": "RRSIG",
                                "type_covered": 1
                            },
                            "status": "Secure"
                        }
                    ],
                    "authoritative": [
                        {
                            "error": "",
                            "rrset": {
                                "class": 1,
                                "name": "example.com.",
                                "type": 2
                            },
                            "sig": {
                                "algorithm": 13,
                                "class": "IN",
                                "expiration": "20241206020003",
                                "inception": "20241115121713",
                                "keytag": 60915,
                                "labels": 2,
                                "name": "example.com",
                                "original_ttl": 86400,
                                "signature": "9W8bAJT/4vA3nH8QU/NeM1n8bZZWn5PBkN26ix827VQGWY0YdMniks/1YCIaVNl16xly4s5IsVDtI9rLvaZORw==",
                                "signer_name": "example.com.",
                                "ttl": 86400,
                                "type": "RRSIG",
                                "type_covered": 2
                            },
                            "status": "Secure"
                        }
                    ],
                    "dnskey": [
                        {
                            "algorithm": 13,
                            "class": "IN",
                            "flags": 256,
                            "name": "example.com",
                            "protocol": 3,
                            "public_key": "ai2pvpijJjeNTpBu4yg6T375JqIStPtLABDTAILb+f4J7XpofUNXGQn6FpQvZ6CARWn2xQapbjGtDRjTf4qYxg==",
                            "ttl": 3600,
                            "type": "DNSKEY"
                        }
                    ],
                    "ds": [
                        {
                            "algorithm": 13,
                            "class": "IN",
                            "digest": "be74359954660069d5c63d200c39f5603827d7dd02b56f120ee9f3a86764247c",
                            "digest_type": 2,
                            "key_tag": 370,
                            "name": "example.com",
                            "ttl": 3600,
                            "type": "DS"
                        }
                    ],
                    "status": "Secure"
                },
                "flags": {
                    "authenticated": false,
                    "authoritative": true,
                    "checking_disabled": false,
                    "error_code": 0,
                    "opcode": 0,
                    "recursion_available": false,
                    "recursion_desired": false,
                    "response": true,
                    "truncated": false
                },
                "protocol": "udp",
            },
            "duration": 1.072630625,
            "status": "NOERROR",
        }
    }
}
```

Example output (normal):
```json
{
    "name": "example.com",
    "results": {
        "A": {
            "data": {
                "answers": [
                    {
                        "answer": "93.184.215.14",
                        "class": "IN",
                        "name": "example.com",
                        "ttl": 3600,
                        "type": "A"
                    },
                    {
                        "algorithm": 13,
                        "class": "IN",
                        "expiration": "20241206041826",
                        "inception": "20241115121713",
                        "keytag": 60915,
                        "labels": 2,
                        "name": "example.com",
                        "original_ttl": 3600,
                        "signature": "vRP5fmIUOIHdIXJsnFkSLYL84OEh4isvOeIkPdGR1Ro+YUs4tTNO6J3ajnDvcIZfc5p+j8R35Z/SOOqD9ZY95w==",
                        "signer_name": "example.com.",
                        "ttl": 3600,
                        "type": "RRSIG",
                        "type_covered": 1
                    }
                ],
                "dnssec": {
                    "status": "Secure"
                },
                "protocol": "udp"
            },
            "duration": 0.803631958,
            "status": "NOERROR"
        }
    }
}
```

- [x] Implement error status handling according to [RFC 4035, Section 4.3](https://www.rfc-editor.org/rfc/rfc4035.html)
- [x] Ensure proper handling of non-secure cases
- [x] Share retries 
- [x] Address all linter issues
- [x] Limit DNSSEC to iterative mode only?
- [x] CLI option

developStorm committed:
refactor: return raw dns response from auxiliary functions

developStorm committed:
feat: basic, same-level DNSSEC validation

developStorm committed:
fix: newLayer should not be updated if extractAuthority fails

developStorm committed:
refactor: extend trace, fix depth for couple places

developStorm committed:
refactor: more comprehensive validation of DNSKEY/RRSIG

dnssec related functions now sit in their own file

developStorm committed:
patch: disregard RRSIG in authorities

developStorm committed:
fix: remove depth parameter in lookup function

revert depth change. depth can't be carried to subqueries

developStorm committed:
feat: implement DS verification

developStorm committed:
feat: cache DS in referrals

developStorm committed:
feat: validate dnssec for referrals

developStorm committed:
build(ci): bump to go1.21

developStorm committed:
fix: answer section is always validated

phillip-stephens commented:
Additionally, let's add integration tests to `integration_tests.py` for positive/negative test cases, as defined here
https://www.internetsociety.org/resources/deploy360/2013/dnssec-test-sites/

developStorm committed:
fix: resolve linter issues

developStorm committed:
Merge branch 'main' into feat/basic-dnssec-validation

developStorm committed:
fix: resolve linter issues

developStorm committed:
refactor: store current retry in resolver struct

phillip-stephens commented on src/zdns/resolver.go:
```diff
@@ -269,8 +269,9 @@ type Resolver struct {
-	retries  int
-	logLevel log.Level
+	retries          int
+	retriesRemaining int
```
Can we add comments that `retries` is the static constant configured on resolver init and `retriesRemaining` is set/decremented for each domain's resolution to make the purpose clear

developStorm replied:
3561c36d1bcf1b0ea914cd31f2777e38b8d362a3

developStorm committed:
patch: don't modify original res

developStorm committed:
docs: rrset validation failure

developStorm committed:
build(deps): bump root anchors

developStorm committed:
fix: add validity period check for RRSIG

developStorm committed:
feat: add dnssec result types

developStorm committed:
docs: clarify comments for retries and retriesRemaining in Resolver

developStorm committed:
feat: dnssec result field

developStorm committed:
refactor: simplify param list with validator struct

developStorm committed:
refactor: extended DNSSEC result handling

developStorm committed:
feat: add JSON tags to DNSSEC types for improved serialization

developStorm committed:
refactor: DNSSEC functions now does not stop resolution

developStorm committed:
fix: regression in authority caching

developStorm committed:
feat: add DNSSEC validation as CLI option

developStorm committed:
feat: implement circular query detection

developStorm commented on src/zdns/conf.go:
```diff
@@ -49,8 +49,17 @@ const (
 
+func isStatusRetryable(status Status) bool {
+	switch status {
+	case StatusServFail, StatusNXDomain, StatusRefused, StatusTruncated, StatusError, StatusTimeout, StatusIterTimeout:
```
@phillip-stephens could you check if these make sense to you

phillip-stephens replied:
Since `retries` are global to the lookup, we shouldn't use a `retry` if there's no chance of getting back a different response.

`NXDomain` we should not retry. Based on [RFC 8020](https://datatracker.ietf.org/doc/html/rfc8020), there's no chance we re-query and get a different response to an NXDomain. So this **should not** be retry-able
The others I'm fine with leaving as-is.

Additionally, let's add `NXDomain` as a specific `case` statement and add that RFC as reason why we will not be re-trying that specific domain for documentation.

And make `StatusTimeout` and `StatusIterTimeout` their own case statement and comment:
`up to caller's implementation to retry` since technically where this function is used in lookup.go, we'll abort the query if the `ctx` expires, the main reason for getting this error.

The others I'm fine with!

developStorm committed:
chore: suppress some lint warnings

developStorm committed:
fix: handle DNSSEC insecure and bogus statuses

developStorm committed:
test: add DNSSEC integration tests

developStorm committed:
fix: RRset should be identified by all of name, class and type

Have this super weird case where additionals from dnssec-tools.org contains an A and RRSIG for each of (nsm|nsw).dnssec-tools.org.
If identify by only type, these two will be clustered under the same set and could not validate.

developStorm committed:
fix: dedup ds/dnskey

developStorm committed:
docs: function comments for dnssec

zakird commented:
What have we done to test this? Can we run this against the top million websites or such and see how it goes? A lot of code.

developStorm commented:
> What have we done to test this? 

At this point it's just manual and unit tests performed against the above domains shared by Phillip, and I monitored request sanity with Wireshark. I can look into large scale test later.

developStorm committed:
fix: shortcut Insecure if entire answer is unsigned

developStorm committed:
feat: add dnssec output fields

developStorm committed:
Revert "test: add DNSSEC integration tests"

This reverts commit 2b074a238585444ab7a3da326962227b15ba4982.

developStorm committed:
test: add DNSSEC integration tests

developStorm committed:
fix: rrsig error handling and DS validation on DNSKEY response

developStorm committed:
fix: KSK is a lie :(

developStorm committed:
feat: lazy query A/AAAA of NSes when iterating on authorities

developStorm committed:
chore: error logging

developStorm committed:
docs: comments

developStorm committed:
fix: skip validation for some sections if answer is authoritative

developStorm committed:
fix: should restore the previous dnssec setting

developStorm committed:
docs: add RFC references

zakird committed:
Merge branch 'main' into feat/basic-dnssec-validation

developStorm committed:
docs: add one rfc reference

developStorm committed:
Merge branch 'main' into feat/basic-dnssec-validation

zakird commented on src/cli/cli.go:
```diff
@@ -67,6 +67,7 @@ type QueryOptions struct {
 	ClassString        string `long:"class" default:"INET" description:"DNS class to query. Options: INET, CSNET, CHAOS, HESIOD, NONE, ANY."`
 	ClientSubnetString string `long:"client-subnet" description:"Client subnet in CIDR format for EDNS0."`
 	Dnssec             bool   `long:"dnssec" description:"Requests DNSSEC records by setting the DNSSEC OK (DO) bit"`
+	ValidateDNSSEC     bool   `long:"validate-dnssec" description:"Validate DNSSEC records, only applicable with --iterative"`
```
I think we discussed allowing this even when not in `--iterative``.

developStorm replied:
DM'ed on this and explained why it's not feasible. If we reach a different conclusion later, happy to implement this in a separate PR.

zakird commented on src/zdns/dnssec_types.go:
```diff
@@ -0,0 +1,160 @@
+	DS            []*DSAnswer          `json:"ds" groups:"dnssec,long,trace"`
+	DNSKEY        []*DNSKEYAnswer      `json:"dnskey" groups:"dnssec,long,trace"`
+	Answer        []DNSSECPerSetResult `json:"answer" groups:"dnssec,long,trace"`
+	Additionals   []DNSSECPerSetResult `json:"additionals" groups:"dnssec,long,trace"`
```
I think this should be `Additional`  not `Additionals` per RFC and in matching the other types.

developStorm replied:
We currently have `json:"additionals` in `SingleQueryResult` 
https://github.com/zmap/zdns/blob/d403578a6ca4aaed047ab6bc452e67b7df4a9419/src/zdns/qa.go#L77-L79

and in `CachedResult`:
https://github.com/zmap/zdns/blob/d403578a6ca4aaed047ab6bc452e67b7df4a9419/src/zdns/cache.go#L35-L38

I can address the pluralization issue for DNSSEC and cache (ideally in a separate PR). However, for `SingleQueryResult`, since the plural form is embedded in the JSON schema, changing it could introduce backward compatibility issues. 

If we leave it unchanged, we’ll have a mismatch where the JSON schema for DNSSEC uses the singular form, but the response retains the plural form. This inconsistency might be confusing.

To resolve this, I suggest we standardize by either:
1. Removing all plural forms and break compatibility, or 
2. Consistently using plural forms, at least for the JSON output.

What do you think?

zakird replied:
We're about to do a major 2.0 release. I think we should have it consistent everywhere, even if it means that schema changes.

developStorm replied:
Fixed in dnssec code. Created #484 for other places.

zakird commented on src/zdns/resolver.go:
```diff
@@ -88,15 +89,16 @@ type ResolverConfig struct {
+	ShouldValidateDNSSEC bool           // whether to validate DNSSEC
+	DNSOverHTTPS         bool           // whether to use DNS over HTTPS for External Lookups, n/a to Iterative Lookups
+	DNSOverTLS           bool           // whether to use DNS over TLS for External Lookups, n/a to Iterative Lookups
+	RootCAs              *x509.CertPool // Root CAs for DoT/DoH Server Verification
```
Do we want to allow people to pass in root store on the command line? Otherwise, where are we expecting it to originate?

developStorm replied:
DNSSEC root anchors are from https://github.com/zmap/go-dns-root-anchors. That repository updates itself as described in RFC7958.

developStorm committed:
refactor: additionals -> additional

developStorm committed:
refactor: passing on validation result from front of the chain

developStorm committed:
Merge branch 'main' into feat/basic-dnssec-validation

developStorm committed:
refactor: miscellaneous bits

developStorm committed:
fix: NSEC3 type definition

developStorm committed:
feat: NSEC3 validation for DS records

developStorm committed:
style: fix linter issues

developStorm committed:
test: add integration test for dnssec with cnames

developStorm committed:
fix: bogus case classification

developStorm committed:
patch: comment

developStorm committed:
feat: support NSEC for DS check

developStorm committed:
fix: always cache under lower-cased authname

developStorm committed:
fix: DS records should always be cached as secure

developStorm committed:
fix: support DS/NSEC in authority section

developStorm committed:
chore: logging for a specific nsec3 case

It seems cloudflare will not validate anything with iteration > 0. We are fine validating them, but print a log for visibility.

developStorm commented:
I conducted another round of tests on the top 10k domains, and the results aligned closely with Cloudflare, with a few explainable non-alignments remain:

1. **Domains under the `.my` TLD**:  
   These cannot be verified by us because they use an oversized RSA exponent that Go does not support. This issue is documented here: https://github.com/golang/go/issues/3161. As discussed with @phillip-stephens, we agree that per RFC 4033, this should be classified as bogus.

2. **Specific domains (`www.correoargentino.com.ar`, `xskt.com.vn`, and `futbollibrehd.pe`)**:  
   Cloudflare flagged these as bogus due to their inability to establish Secure Entry Point (SEP). Upon reviewing [DNSViz](https://dnsviz.net/d/xskt.com.vn/dnssec/), it appears these domains use NSEC3 records with iterations > 0, which is non-compliant with [RFC9276](https://datatracker.ietf.org/doc/html/rfc9276#section-3.1). Cloudflare likely ignored these NSEC3 records, rendering them unable to confirm the non-existence of DS records.  

   Since we are not a publicly facing recursive resolver and are not exposed to the risk of DoS attacks from this, I decided that we would continue to validate these NSEC3 records while printing a log: 9f7d8913ef1e5af17151c80f74d2c26ab4034452.

I'm comfortable with the shape of this PR now. Feel free to start reviewing @phillip-stephens @zakird.

zakird reviewed (approved):
Overall this look great and I'm very excited to see DNSSEC validation added. @phillip-stephens is going to take a closer look.

zakird commented:
Re 1: This might be worth fixing in ZCrypto and using that. I think that we can leave this however as a FR and merge in any case.

phillip-stephens committed:
revert change to go 1.21.1

phillip-stephens committed:
downgrade required go version to 1.20

phillip-stephens reviewed (approved):
LGTM, I tested it locally on 1k domains and spot-checked a few of them. Also reverted the change to require `go 1.20` since not sure we need to do that. Otherwise this looks really great and super excited about it! Thanks @developStorm!

phillip-stephens committed:
Merge branch 'main' into feat/basic-dnssec-validation

phillip-stephens committed:
fix lint error

phillip-stephens committed:
toolchain directive not available in go 1.20

phillip-stephens committed:
upgrade -> go 1.21
